src/zenml/integrations/mlflow/steps/mlflow_deployer.py

```diff
@@ -111,98 +111,83 @@ def mlflow_model_deployer_step(
-        mlserver=mlserver,
-        pipeline_name=pipeline_name,
-        run_name=run_name,
-        pipeline_run_id=run_name,
-        pipeline_step_name=step_name,
-        timeout=timeout,
-    )
-
-    # Creating a new service with inactive state and status by default
-    service = MLFlowDeploymentService(predictor_cfg)
-    if existing_services:
-        service = cast(MLFlowDeploymentService, existing_services[0])
-
-    # check for conditions to deploy the model
-    if not model_uri:
-        # an MLflow model was not trained in the current run, so we simply reuse
-        # the currently running service created for the same model, if any
-        if not existing_services:
-            logger.warning(
-                f"An MLflow model with name `{model_name}` was not "
-                f"logged in the current pipeline run and no running MLflow "
-                f"model server was found. Please ensure that your pipeline "
-                f"includes a step with a MLflow experiment configured that "
-                "trains a model and logs it to MLflow. This could also happen "
-                "if the current pipeline run did not log an MLflow model  "
-                f"because the training step was cached."
-            )
-            # return an inactive service just because we have to return
-            # something
-            return service
+    # Check whether to deploy a new service
+    if model_uri and deploy_decision:
+        predictor_cfg = MLFlowDeploymentConfig(
+            model_name=model_name or "",
+            model_uri=model_uri,
+            workers=workers,
+            mlserver=mlserver,
+            pipeline_name=pipeline_name,
+            run_name=run_name,
+            pipeline_run_id=run_name,
+            pipeline_step_name=step_name,
+            timeout=timeout,
+        )
+        new_service = cast(
+            MLFlowDeploymentService,
+            model_deployer.deploy_model(
+                replace=True,
+                config=predictor_cfg,
+                timeout=timeout,
+            ),
+        )
         logger.info(
-            f"An MLflow model with name `{model_name}` was not "
-            f"trained in the current pipeline run. Reusing the existing "
-            f"MLflow model server."
+            f"MLflow deployment service started and reachable at:\n"
+            f"    {new_service.prediction_url}\n"
         )
-        if not service.is_running:
-            service.start(timeout)
-
-        # return the existing service
-        return service
-
-    # even when the deploy decision is negative, if an existing model server
-    # is not running for this pipeline/step, we still have to serve the
-    # current model, to ensure that a model server is available at all times
-    if not deploy_decision and existing_services:
+        logger.info("Stopping existing services...")
+        for existing_service in existing_services:
+            existing_service.stop(timeout)
```

Comment:
Major question: why not reuse? I would search existing by Model Name and Model Version and if same - just reuse.

Reply:
But why do we deploy, if we find the existing model with the same version? It is the same model - nothing changed.
It is related directly to the previous question.

Now we identify the Server by pipeline, step, and model name and this is probably a reason, cause we have no clue if the model actually changed or not, but if we switch to name+version, there would be no good reason to redeploy the same version, imho.

Reply:
Actually model version only exists for model registries right? In this specific step there might not be a model registry so models might not be versioned at all.

Reply:
But thanks for mentioning this, I almost forgot to adjust the model registry deployer step as well